### PR TITLE
[WIP] Add perspective camera for td viewport

### LINKED ISSUE #6260: Provide Perspective Camera for 3D viewport

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
This pr tries to add a perspective camera to the 3D viewport. More information is posted below ⬇️ 

### URL of deployed dev instance (used for testing):
- https://addperspectivecamerafortd.webknossos.xyz

### Steps to test:
- visit https://addperspectivecamerafortd.webknossos.xyz/datasets/sample_organization/l4_sample/view#3584,3584,1024,0,1.3
- (log in)
- use the 3D viewport and ensure that the perspective camera setting is enabled (see screenshot)
image
- pan or rotate with left- or right-click drag
- both panning and rotating should behave natural, but feel buggy currently (the rotation center should always be the intersection of the visible planes; see screenshot)
image
- zoom in/out is done via mouse wheel

### Issues:
- fixes #6260

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Needs datastore update after deployment
- [ ] Ready for review
-----

image

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
What I noticed is that our version of the trackball we use is outdated compared to the current version of the trackball of threejs. The [current version of the threejs trackball](https://github.com/mrdoob/three.js/blob/dev/examples/js/controls/TrackballControls.js) includes quite some if else statements depending on the type of the camera. This might hint at what out trackball is doing wrong for the perspective camera and maybe also why the rotating is messed up, when the flycam position is not near the center of the 3d viewport?

^ this was just a random thought of mine that might be helpful

File: frontend/javascripts/oxalis/view/plane_view.ts [hunk 1]
```diff
@@ -40,46 +50,63 @@ class PlaneView {
...
+
+    for (const plane of OrthoViewValuesWithoutTDView) {
+      // Let's set up cameras
+      partialCameraMap[plane] = getNewCamera(plane, true);
     }
+    partialCameraMap[OrthoViews.TDView] = {
+      [TDCameras.OrthographicCamera]: getNewCamera(`${OrthoViews.TDView}_Orthographic`, true),
+      [TDCameras.PerspectiveCamera]: getNewCamera(`${OrthoViews.TDView}_Perspective`, false),
+    };
+    const cameras: OrthoViewCameraMap = partialCameraMap;
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/oxalis/view/plane_view.ts:94]: The workaround `partialCameraMap`is needed to get the cameras into a single object. Other approaches lead to typescript errors.

**Comment by MichaelBuessemeyer** [comment]
Also: note that I just pushed the newest version of the trackball of the threejs repo converted to typescript to this pr. I removed it again as it is currently not used. I did not replace the old trackball with it, as the new version uses other events as our old version and the new events sadly also trigger when clicking one of the td viewport buttons which lead to weird and unwanted behaviour when I wanted to switch to the new trackball version. But the weird behaviour might have also been caused by the unstable state this branch was in during that time. Maybe it is possible to use the new version now 🤔 . (<- this is just another of my side thoughts).

I removed the new version of the trackball directly in the next commit as it is not needed in the current version of the pr. Also, typescript is unhappy about some typing errors with the current stage of this pr. Nevertheless, the new trackball is persisted in the commit history thanks to the power of git. (and this was the main goal of these two commits :))

File: frontend/javascripts/oxalis/constants.ts [hunk 2]
```diff
@@ -49,17 +50,30 @@ export enum OrthoViews {
...
   PLANE_XY = "XY",
   PLANE_YZ = "YZ",
   PLANE_XZ = "XZ",
   TDView = "3D",
 }
 export type OrthoView = keyof typeof OrthoViews;
-export type OrthoViewWithoutTD = Exclude<keyof typeof OrthoViews, OrthoViews.TDView>;
-
+// TODO: typing was broken here, find a better workaround than using "TDView" directly.
+export type OrthoViewWithoutTD = Exclude<keyof typeof OrthoViews, "TDView">;
```
[the remark this answers is not in the document]
Comment (philippotto) [reply frontend/javascripts/oxalis/constants.ts:65]: what a coincidence. I found the same issue today too ^^ I also changed it to `"TDView"` (on the webgl2 branch) and think that this is acceptable.

**Comment by philippotto** [comment]
> Also: note that I just pushed the newest version of the trackball of the threejs repo converted to typescript to this pr. I removed it again as it is currently not used. I did not replace the old trackball with it, as the new version uses other events as our old version and the new events sadly also trigger when clicking one of the td viewport buttons which lead to weird and unwanted behaviour when I wanted to switch to the new trackball version. But the weird behaviour might have also been caused by the unstable state this branch was in during that time. Maybe it is possible to use the new version now  .

The conversion to typescript wasn't done by you, right? I think, it would make sense to use the newest trackball code to reduce technical debt and potentially piggyback on bug fixes somebody else did.

The top priority for the bug fixes should be (1) getting panning and rotating right (via mouse). Afterwards, (2) the rotation buttons should be fixed. The last step would be (3) perfectly synchronizing perspective and orthographic camera.
Doing (1) is probably the most difficult, but also the most important one. And (3) doesn't seem super critical to me (as this won't be a frequent user action).
If using the newest trackball code helps for (1), I'd do so. If it doesn't help, we at least tried it out.

**Comment by MichaelBuessemeyer** [comment]
> The conversion to typescript wasn't done by you, right?

It was done by me. It took some time but my IDE (vscode) was pretty helpful in that regard 🎉

File: frontend/javascripts/oxalis/constants.ts [hunk 3]
```diff
@@ -89,7 +103,7 @@ export const OrthoViewValuesWithoutTDView = [
   OrthoViews.PLANE_XY,
   OrthoViews.PLANE_YZ,
   OrthoViews.PLANE_XZ,
-];
+] as [OrthoViews.PLANE_XY, OrthoViews.PLANE_YZ, OrthoViews.PLANE_XZ];
```
Comment (philippotto) [inline frontend/javascripts/oxalis/constants.ts:106]:
```suggestion
] as const;
```

**might** work too.

**Comment by philippotto** [comment]
> > The conversion to typescript wasn't done by you, right?
> 
> It was done by me. It took some time but my IDE (vscode) was pretty helpful in that regard tada

wow, cool :+1:  then it would be a pity to throw it away :)

File: frontend/javascripts/libs/trackball_controls.ts [hunk 4]
```diff
@@ -211,9 +211,12 @@ const TrackballControls = function (
         angle *= _this.rotateSpeed;
         quaternion.setFromAxisAngle(axis, -angle);
 
-        _eye.applyQuaternion(quaternion);
+        _eye.sub(_this.flycamPos).applyQuaternion(quaternion).add(_this.flycamPos);
 
         _this.object.up.applyQuaternion(quaternion);
+        _this.object.position.sub(_this.flycamPos).applyQuaternion(quaternion).add(_this.flycamPos);
+
+        _this.target.sub(_this.flycamPos).applyQuaternion(quaternion).add(_this.flycamPos);
```
Comment (philippotto) [inline frontend/javascripts/libs/trackball_controls.ts:219]: The rotation only works if the flycam position is centered in the 3D viewport. If the user pans a bit and then rotates, the rotation goes wild. The code here might be responsible for this.

**Comment by philippotto** [comment]
Closing for now due prioritization.

**Pull request closed without merging**